plugins/inputs/apcupsd/apcupsd.go

```diff
@@ -0,0 +1,75 @@
+package apcupsd
+
+import (
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/plugins/inputs"
+	"github.com/mdlayher/apcupsd"
+)
+
+const defaultAddress = "127.0.0.1:3551"
+
+type ApcUpsd struct {
+	Servers []string
+}
+
+func (_ *ApcUpsd) Description() string {
+	return "Monitor APC UPSes connected to apcupsd"
+}
+
+var sampleConfig = `
+  # a list of running apcupsd server to connect to. 
+  # If not provided will default to 127.0.0.1:3551
+  servers = ["127.0.0.1:3551"]
+`
+
+func (_ *ApcUpsd) SampleConfig() string {
+	return sampleConfig
+}
+
+func (h *ApcUpsd) Gather(acc telegraf.Accumulator) error {
+	for _, addr := range h.Servers {
+		status, err := fetchStatus(addr)
+		if err != nil {
+			return err
+		}
+
+		tags := map[string]string{
+			"serial":   status.SerialNumber,
+			"ups_name": status.UPSName,
+		}
+
+		fields := map[string]interface{}{
+			"online":                 status.Status == "ONLINE",
+			"input_voltage":          status.LineVoltage,
+			"load_percent":           status.LoadPercent,
+			"battery_charge_percent": status.BatteryChargePercent,
+			"time_left":              status.TimeLeft.Minutes(),
+			"output_voltage":         status.OutputVoltage,
+			"internal_temp":          status.InternalTemp,
+			"battery_voltage":        status.BatteryVoltage,
+			"input_frequency":        status.LineFrequency,
+			"time_on_battery":        status.TimeOnBattery.Minutes(),
+		}
+
+		acc.AddFields("apcupsd", fields, tags)
+	}
+	return nil
+}
+
+func fetchStatus(addr string) (*apcupsd.Status, error) {
+	client, err := apcupsd.Dial("tcp", addr)
```

Comment:
Would it be possible to add timeouts on all the network operations, perhaps using a `Context`?  This would require a modification in the upstream library but would prevent network issues from making Telegraf unresponsive.

Reply:
Should that timeout be configurable in the telegraf config or should i just set it to like 5 sec?